### PR TITLE
Only clear GL context after changing the thread configuration

### PR DESCRIPTION
**Pull request by blasten**
## Description

The GL context must be clear in the current thread prior to making it current in a different thread. In this PR, the thread merger callback can be used to perform such clean ups.  This is a more conservative approach than always clearing the context.

## Tests

I added the following tests: Unit tests

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Review by jason-simmons — APPROVED** [review]

**Comment by flar** [comment]
I'm seeing this error on both the Linux and Mac Host Engine tasks:

```
[ RUN      ] ShellTest.OnPlatformViewDestroyAfterMergingThreads
[FATAL:flutter/fml/memory/weak_ptr.h(177)] Check failed: (checker_.checker).RunsOnCreationTaskRunner(). 
```

**Comment by blasten** [comment]
@flar should be good now. I'm going to run the firebase tests for 30 min to ensure there's no crash.

File: shell/common/rasterizer.cc [hunk 1]
```diff
@@ -89,6 +89,14 @@ void Rasterizer::Setup(std::unique_ptr<Surface> surface) {
         delegate_.GetTaskRunners().GetRasterTaskRunner()->GetTaskQueueId();
     raster_thread_merger_ =
         fml::MakeRefCounted<fml::RasterThreadMerger>(platform_id, gpu_id);
+
+    raster_thread_merger_->SetMergeUnmergeCallback(
```
Comment (jason-simmons) [inline shell/common/rasterizer.cc]: Does the merge/unmerge callback need to be removed during rasterizer teardown?

The callback is holding a pointer to the rasterizer, so the rasterizer needs to ensure that the callback does not outlive the rasterizer.

Comment (blasten) [reply shell/common/rasterizer.cc]: Done in https://github.com/flutter/engine/pull/20965/commits/2893c3434cc1498bd592154a7fb7def9efe5e90c

**Pull request merged**